sdk/servicebus/azure-servicebus/azure/servicebus/aio/_servicebus_receiver_async.py

```diff
@@ -212,40 +212,38 @@ def __init__(
             None if self._session_id is None else ServiceBusSession(cast(str, self._session_id), self)
         )
         self._receive_context = asyncio.Event()
-        self._handler: ReceiveClientAsync
-
-    # Python 3.5 does not allow for yielding from a coroutine, so instead of the try-finally functional wrapper
-    # trick to restore the timeout, let's use a wrapper class to maintain the override that may be specified.
-    class _IterContextualWrapper(collections.abc.AsyncIterator):
-        def __init__(self, receiver, max_wait_time=None):
-            self.receiver = receiver
-            self.max_wait_time = max_wait_time
-
-        async def __anext__(self):
-            # pylint: disable=protected-access
-            original_timeout = None
+
+    async def _iter_contextual_wrapper(self, max_wait_time=None):
+        # pylint: disable=protected-access
+        original_timeout = None
+        while True:
+            print("In iter")
             # This is not threadsafe, but gives us a way to handle if someone passes
             # different max_wait_times to different iterators and uses them in concert.
-            if self.max_wait_time and self.receiver and self.receiver._handler:
-                # TODO: What did the previous _handler.timeout represent here?
-                original_timeout = self.receiver._handler._idle_timeout
-                self.receiver._handler._timeout = self.max_wait_time * 1000
+            if max_wait_time:
+                original_timeout = self._handler._timeout
+                self._handler._timeout = max_wait_time * 1
             try:
-                self.receiver._receive_context.set()
-                message = await self.receiver._inner_anext()
+                # self._receive_context.set()
+                message = await self._inner_anext()
                 links = get_receive_links(message)
-                with receive_trace_context_manager(self.receiver, links=links):
-                    return message
+                with receive_trace_context_manager(self, links=links):
+                    yield message
+                    # print("Yielded")
+            except StopAsyncIteration:
+                print("Stop Iteration")
+                break
             finally:
-                self.receiver._receive_context.clear()
```

Comment:
I am wondering if this is why they had the 3 while loops because other wise on the async side you hit this finally block for the previous iterator once the current has started and it messes up the receive_context being set so you end up releasing messages you want to keep

Reply:
acutally I think this is a test-side issue, but I am wondering if on the iterator some exceptions from the tests sending received messages are being hidden